tools/lsp/completion.rs

```diff
@@ -82,6 +86,105 @@ pub(crate) fn completion_at(
                     with_insert_text(c, ins_tex, client_caps)
                 }),
             );
+
+            // Find out types that can be imported
+            let import_locations = Some(()).filter(|()| snippet_support).and_then(|()| {
+                let current_file = token.source_file.path().to_owned();
+                let current_doc =
+                    document_cache.documents.get_document(&current_file)?.node.as_ref()?;
+                let current_uri = lsp_types::Url::from_file_path(&current_file).ok()?;
+                let mut import_locations = HashMap::new();
+                let mut last = 0u32;
+                for import in current_doc.ImportSpecifier() {
+                    if let Some((loc, file)) = import.ImportIdentifierList().and_then(|list| {
+                        Some((
+                            document_cache.byte_offset_to_position(
+                                list.ImportIdentifier().last()?.text_range().end().into(),
+                                &current_uri,
+                            )?,
+                            import.child_token(SyntaxKind::StringLiteral)?,
+                        ))
+                    }) {
+                        import_locations
+                            .insert(file.text().to_string().trim_matches('\"').to_string(), loc);
+                    }
+                    last = import.text_range().end().into();
+                }
+                let last = if last == 0 {
+                    0
+                } else {
+                    document_cache
+                        .byte_offset_to_position(last, &current_uri)
+                        .map_or(0, |p| p.line + 1)
+                };
+                Some((import_locations, last, current_uri))
+            });
+
+            if let Some((import_locations, last, current_uri)) = import_locations {
+                for file in document_cache.documents.all_files() {
+                    let doc = document_cache.documents.get_document(file).unwrap();
+                    let file = if file.starts_with("builtin:/") {
+                        match file.file_name() {
+                            Some(file) if file == "std-widgets.slint" => "std-widgets.slint".into(),
+                            _ => continue,
+                        }
+                    } else {
+                        match lsp_types::Url::make_relative(
+                            &current_uri,
+                            &lsp_types::Url::from_file_path(dbg!(file)).unwrap(),
+                        ) {
+                            Some(file) => file,
+                            None => continue,
+                        }
+                    };
+
+                    for (exported_name, ty) in &doc.exports.0 {
```

Comment:
I think this would read nicer if exports deref’d to the inner, so that the .0 isn’t required here.